semantic/src/main/kotlin/tools/samt/semantic/checks/ValidTypeExpressionCheck.kt

```diff
@@ -0,0 +1,155 @@
+package tools.samt.semantic.checks
+
+import tools.samt.common.DiagnosticConsole
+import tools.samt.parser.*
+import tools.samt.semantic.SemanticCheck
+
+internal class ValidTypeExpressionCheck(private val diagnostics: DiagnosticConsole) : SemanticCheck {
+    override fun check(fileNode: FileNode) {
+        fileNode.statements.forEach {
+            when (it) {
+                is RecordDeclarationNode -> it.fields.forEach { field -> checkTypeExpression(field.type) }
+                is TypeAliasNode -> checkTypeExpression(it.type)
+                is TypeImportNode,
+                is WildcardImportNode,
+                is EnumDeclarationNode,
+                is ConsumerDeclarationNode,
+                is PackageDeclarationNode,
+                is ProviderDeclarationNode,
+                is ServiceDeclarationNode,
+                -> Unit
+            }
+        }
+    }
+
+    private fun checkTypeExpression(expression: ExpressionNode, optionalAllowed: Boolean = true) {
+        when (expression) {
+            is IdentifierNode,
+            is BundleIdentifierNode,
+            -> Unit
+
+            is OptionalDeclarationNode -> {
```

Comment:
I think it makes more sense to perform this check once we have type information available. That way we can catch all cases of nested optionals (since right now we don't know wether an Identifier refers to an optional type already). By deferring this check until later we don't perform the check twice and have all the nested optional checking on one place.

Reply:
Well the question is, should an optional for an already optional parameter be an error? We could also handle it like TypeScript and just accepting an infinite amount of `? ` if the user so desires 🤔 

I think the following code should work without any issue, do you disagree?
```
alias A: String?
alias B: A?
```

Reply:
I agree that it shouldn't be an error, but a warning (e.g `warning: type is already declared as optional, duplicate optional declaration is ignored`) would be nice.

Reply:
Should I leave the optional checking as-is for the time being and adjust it once type-checking exists or remove the code for now?

Reply:
Yes leave it in for the moment